### PR TITLE
[NOT YET] All of Leo: code review

### PR DESCRIPTION
**Pull request by helgridly**
Once we've got Leo working end-to-end I'd like us to swarm on a code review and do some housekeeping tasks. This PR tracks the current state of develop against the beginning of time so it contains everything we've done.

code review date TBD but not yet

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+- [ ] **Submitter**: Include the JIRA issue number in the PR description
```
Comment (helgridly) [inline .github/PULL_REQUEST_TEMPLATE.md]: switch this with slimmer PR checkbox list from orch

File: jupyter-docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+ENV JAVA_URL http://download.oracle.com/otn-pub/java/jdk/8u131-b11/d54c1d3a095b4ff2b6607d096fa80163/$JAVA_TGZ
+ENV JAVA_HOME /usr/lib/jdk/$JAVA_VER
+
+ENV SPARK_VER 2.0.2
+ENV SPARK_TGZ spark-$SPARK_VER.tgz
+ENV SPARK_URL http://d3kbcqa49mib13.cloudfront.net/$SPARK_TGZ
+ENV SPARK_HOME /spark/spark-$SPARK_VER
+
+# result of `gsutil cat gs://hail-common/latest-hash.txt` on 6 June 2017
+ENV HAILHASH 53e9d33
```
Comment (helgridly) [inline jupyter-docker/Dockerfile]: Q: as the dev team responsible for maintaining this dependency, how do we do a good job of keeping this up to date and knowing it needs bumping?

File: project/Merging.scala [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import sbtassembly.{MergeStrategy, PathList}
+
+object Merging {
+  def customMergeStrategy(oldStrategy: (String) => MergeStrategy):(String => MergeStrategy) = {
+    case PathList("org", "joda", "time", "base", "BaseDateTime.class") => MergeStrategy.first
```
Comment (helgridly) [inline project/Merging.scala]: can this go? we don't use joda, right?

File: src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changelog.xml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.4.xsd">
+    <include file="changesets/cluster.xml" relativeToChangelogFile="true"/>
+    <include file="changesets/label.xml" relativeToChangelogFile="true"/>
+    <include file="changesets/cluster-unique-key.xml" relativeToChangelogFile="true"/>
+    <include file="changesets/cluster-googleId-unique-key.xml" relativeToChangelogFile="true"/>
+    <include file="changesets/cluster-createdDate-default.xml" relativeToChangelogFile="true" />
```
Comment (helgridly) [inline src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changelog.xml]: i've been wondering if we should date these like we do in rawls.

File: docker/build.sh [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+PROJECT=leonardo
+
+function make_jar()
+{
+    echo "building jar..."
+    # start test db
+    bash ./docker/run-mysql.sh start ${PROJECT}
+
+    # Get the last commit hash of the model directory and set it as an environment variable
+    GIT_MODEL_HASH=$(git log -n 1 --pretty=format:%h)
```
Comment (helgridly) [inline docker/build.sh]: this stuff can go, we don't use it

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      .setServiceAccountPrivateKeyFromPemFile(new java.io.File(dataprocConfig.serviceAccountPemPath))
+      .build()
+  }
+
+
+  def createCluster(googleProject: String, clusterName: String, clusterRequest: ClusterRequest)(implicit executionContext: ExecutionContext): Future[ClusterResponse] = {
+    val op = build(googleProject, clusterName, clusterRequest)
+    op.map{op =>
+      val metadata = op.getMetadata
+      ClusterResponse(clusterName, googleProject, metadata.get("clusterUuid").toString, metadata.get("status").toString, metadata.get("description").toString, op.getName)}
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala]: nit: tidy newlines and spacing here

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    Future {
+      //currently, the bucketPath of the clusterRequest are not used - it will be used later as a place to store notebooks and results
+      val metadata = clusterRequest.labels + ("docker-image" -> dataprocConfig.dataprocDockerImage)
+      val gce = new GceClusterConfig().setMetadata(metadata.asJava).setServiceAccount(clusterRequest.serviceAccount)
+      val initActions = Seq(new NodeInitializationAction().setExecutableFile(dataprocConfig.dataprocInitScriptURI))
+      val clusterConfig = new ClusterConfig().setGceClusterConfig(gce).setInitializationActions(initActions.asJava)
+      val cluster = new Cluster().setClusterName(clusterName).setConfig(clusterConfig)
+      val request = dataproc.projects().regions().clusters().create(googleProject, dataprocConfig.dataprocDefaultZone, cluster)
+      try {
+        executeGoogleRequest(request)
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala]: switch our usages of `executeGoogleRequest` to something like `retryWhen500OrGoogleError`

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+
+  def createCluster(googleProject: String, clusterName: String, clusterRequest: ClusterRequest)(implicit executionContext: ExecutionContext): Future[ClusterResponse] = {
+    val op = build(googleProject, clusterName, clusterRequest)
+    op.map{op =>
+      val metadata = op.getMetadata
+      ClusterResponse(clusterName, googleProject, metadata.get("clusterUuid").toString, metadata.get("status").toString, metadata.get("description").toString, op.getName)}
+  }
+
+  private def build(googleProject: String, clusterName: String, clusterRequest: ClusterRequest)(implicit executionContext: ExecutionContext): Future[Operation] = {
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala]: rename?

Comment (helgridly) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala]: or just merge into `createCluster`

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+      val metadata = clusterRequest.labels + ("docker-image" -> dataprocConfig.dataprocDockerImage)
+      val gce = new GceClusterConfig().setMetadata(metadata.asJava).setServiceAccount(clusterRequest.serviceAccount)
+      val initActions = Seq(new NodeInitializationAction().setExecutableFile(dataprocConfig.dataprocInitScriptURI))
+      val clusterConfig = new ClusterConfig().setGceClusterConfig(gce).setInitializationActions(initActions.asJava)
+      val cluster = new Cluster().setClusterName(clusterName).setConfig(clusterConfig)
+      val request = dataproc.projects().regions().clusters().create(googleProject, dataprocConfig.dataprocDefaultZone, cluster)
+      try {
+        executeGoogleRequest(request)
+      } catch {
+        case e: GoogleJsonResponseException => throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"Google Request Failed: ${e.getDetails.getMessage}"))
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala]: kill `WorkbenchExceptionWithErrorReport` with extreme prejudice.

also decide how to handle google errors and how much to tell users. jeff suggests from cromwell experience that if we don't know how to handle an error and have given up retrying, pass the error message and HTTP status code up to the user

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/LeoRoutes.scala [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+        }
+      }
+  }
+  def route: Route = (logRequestResult & handleExceptions(myExceptionHandler)) {
+    swaggerRoutes ~ unauthedRoutes ~
+    pathPrefix("api") { leoRoutes ~ proxyRoutes }
+  }
+
+  private val myExceptionHandler = {
+    ExceptionHandler {
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/LeoRoutes.scala]: we should be logging the exceptions thrown here, including their stacktraces

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/ProxyService.scala [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+
+    // Note: we're calling toStrict here which forces the proxy to load the entire response into memory before
+    // returning it to the caller. Technically, this should not be required: we should be able to stream all
+    // data between client, proxy, and server. However Jupyter is doing something strange and the proxy only
+    // works when toStrict is used. Luckily, it's only needed for HTTP requests (which are fairly small) and not
+    // WebSocket requests (which could potentially be large).
+    val handler: Future[HttpResponse] = Source.single(newRequest)
+      .via(flow)
+      .runWith(Sink.head)
+      .flatMap(_.toStrict(5 seconds))
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/ProxyService.scala]: should this 5 seconds be configurable?

**Comment by rtitle** [comment]
ok but we're not done yet! there are still some changes to get it. Maybe we should hold off on the mega review.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+import org.broadinstitute.dsde.workbench.leonardo.model.{Cluster, ClusterRequest, ClusterResponse, ClusterStatus, LeoException}
+import org.broadinstitute.dsde.workbench.leonardo.db.{DataAccess, DbReference}
+import org.broadinstitute.dsde.workbench.leonardo.model.ModelTypes.GoogleProject
+import slick.dbio.DBIO
+
+import scala.concurrent.{ExecutionContext, Future}
+
+case class ClusterNotFoundException(googleProject: GoogleProject, clusterName: String) extends LeoException(s"Cluster $googleProject/$clusterName not found", StatusCodes.NotFound)
+
+class LeonardoService(gdDAO: DataprocDAO, dbRef: DbReference)(implicit val executionContext: ExecutionContext) extends LazyLogging {
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: our class names are inconsistent. we should decide whether things are called `LeoFoo` or `LeonardoFoo`. i vote `LeoFoo` but we keep the package and repo named `leonardo`

**Comment by rtitle** [comment]
Looking at the commit list I see a lot of non-squashed commits 😠

**Comment by helgridly** [comment]
@rtitle yeah, i'm just getting a headstart, mainly so i don't forget. this was supposed to be a secret :)

**Comment by rtitle** [comment]
ok i'll shut up :)

File: jupyter-docker/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+ENV DEBIAN_REPO http://cdn-fastly.deb.debian.org
+
+ENV JAVA_VER jdk1.8.0_131
+ENV JAVA_TGZ jdk-8u131-linux-x64.tar.gz
+ENV JAVA_URL http://download.oracle.com/otn-pub/java/jdk/8u131-b11/d54c1d3a095b4ff2b6607d096fa80163/$JAVA_TGZ
+ENV JAVA_HOME /usr/lib/jdk/$JAVA_VER
+
+ENV SPARK_VER 2.0.2
+ENV SPARK_TGZ spark-$SPARK_VER.tgz
+ENV SPARK_URL http://d3kbcqa49mib13.cloudfront.net/$SPARK_TGZ
```
Comment (helgridly) [inline jupyter-docker/Dockerfile]: this looks like a totally legit domain name and not worrying at all

**Pull request closed without merging**